[JENKINS-31458] Call Descriptor.newInstance even on nested data-bound objects

jglick opened the pull request:
[JENKINS-31458](https://issues.jenkins-ci.org/browse/JENKINS-31458)
#1888 introduced a regression for `ParameterDefinition` implementations with customized `newInstance` overrides. This could probably have been fixed by restoring the old `ParametersDefinitionProperty.DescriptorImpl.newInstance` override, which (indirectly) called `newInstance` on the `ParameterDescriptor`s.

However this is merely one case of a longstanding design flaw in Jenkins data binding, that Stapler knows nothing about `Descriptor.newInstance` and so any calls to `newInstance` on a “top-level” object (one instantiated directly by Jenkins code) will not in turn pay attention to custom `newInstance` implementations in nested objects. I have noticed this problem repeatedly in `workflow-plugin`, where I often need to use configuration snippets of `Describable`s at a different level of nesting than Jenkins would normally use.

The fix is to use the existing but rarely implemented `BindInterceptor` hook in Stapler to override the creation of nested `Describable`s to call `newInstance`. In most cases that uses the default implementation (perhaps called as a `super` method from an override), which just asks Stapler to do the work, so we need per-request state to track which JSON fragments we have already encountered, to avoid a `StackOverflowError`.

@reviewbybees esp. @kohsuke since this is pretty fundamental.

jglick committed:
[JENKINS-31458] Reproduced problem in test (passes in 53c86ec but fails in 600b1f0).

jglick committed:
[FIXED JENKINS-31458] Ensuring that Descriptor.newInstance is called even on nested objects.

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

jglick committed:
newInstance may be passed a null request argument.

oleg-nenashev commented on core/src/main/java/hudson/model/Descriptor.java:
```diff
@@ -588,6 +597,65 @@ public T newInstance(StaplerRequest req, JSONObject formData) throws FormExcepti
+            try {
+                return Jenkins.getActiveInstance().getDescriptor(actualType).newInstance(Stapler.getCurrentRequest(), json);
+            } catch (Exception x) {
+                LOGGER.log(Level.WARNING, "falling back to default instantiation " + actualType + " " + json, x);
```
Also catches IllegalStateException from getActiveInstance(). Maybe OK

jglick replied:
Yeah, I do not see a problem with falling back to the previous behavior in that case. Anyway no form binding should be taking place during shutdown.

oleg-nenashev commented on core/src/main/java/hudson/model/Descriptor.java:
```diff
@@ -588,6 +597,65 @@ public T newInstance(StaplerRequest req, JSONObject formData) throws FormExcepti
+
+        @Override
+        public Object onConvert(Type targetType, Class targetTypeErasure, Object jsonSource) {
+            if (jsonSource instanceof JSONObject) {
```
What about JSON Arrays? Seems to be a valid case

jglick replied:
In that case the Stapler methods ought to eventually call `onConvert` with the element type and each of the array elements, so we will intercept it then.

oleg-nenashev commented:
newInstance() is the deprecated approach AFAIK, but it's still popular in plugins. So the change looks to be reasonable to me.

jglick commented:
> newInstance() is the deprecated approach AFAIK

Well, it is only deprecated to the extent that you can use a purely conventional `@DataBoundConstructor` (optionally plus `@DataBoundSetter`s) instead. But there are plenty of cases where that is not an option.

jglick committed:
NodePropertyTest.invisibleProperty demonstrates that the fallback for onConvert must be onConvert, not instantiate, of the delegate.

jglick commented:
Found a problem with `RateLimitBranchProperty.JobPropertyImpl.Descriptor.newInstance`, which calls `super`.

jglick committed:
RateLimitBranchProperty.JobPropertyImpl.Descriptor.newInstance called super.
This caused a NullPointerException since newInstance was then called reëntrantly.
Adjusting NewInstanceBindInterceptor.processed handling to behave correctly in this case.

amuniz commented on core/src/main/java/hudson/model/Descriptor.java:
```diff
@@ -574,7 +575,20 @@ public T newInstance(StaplerRequest req, JSONObject formData) throws FormExcepti
+                BindInterceptor oldInterceptor = req.getBindInterceptor();
+                try {
+                    NewInstanceBindInterceptor interceptor;
+                    if ((oldInterceptor instanceof NewInstanceBindInterceptor)) {
```
Additional `(` and `)`

jglick replied:
True, left over from a refactoring of some `if` clauses.

daniel-beck commented on core/src/main/java/hudson/model/Descriptor.java:
```diff
@@ -588,6 +602,78 @@ public T newInstance(StaplerRequest req, JSONObject formData) throws FormExcepti
+    private static class NewInstanceBindInterceptor extends BindInterceptor {
+
+        private final BindInterceptor oldInterceptor;
+        private final Map<JSONObject,Void> processed = new IdentityHashMap<>();
```
Not clear what this does or why it's reset before the `verifyNewInstance` call.

jglick replied:
> Not clear what this does

Prevents a `StackOverflowError`. Noted in PR description.



> why it's reset before the `verifyNewInstance` call

It is not reset, what do you mean? The same `NewInstanceBindInterceptor` should be used for the duration of the `StaplerRequest` (except when `Node.reconfigure` is called, which causes another one to be used, but that should be harmless).

daniel-beck commented:
Will all of this logging of `JSONObject` result in plain logging of sensitive data, when forms contain password fields, for example?

daniel-beck commented on test/src/test/java/hudson/model/JobPropertyTest.java:
```diff
@@ -115,7 +133,13 @@ public JobPropertyWithConfigImpl(String name) {
-        public static class DescriptorImpl extends JobPropertyDescriptor {}
+        public static class DescriptorImpl extends JobPropertyDescriptor {
+            @Override
+            public JobProperty<?> newInstance(StaplerRequest req, JSONObject formData) throws FormException {
```
Shouldn't this be a separate test? Looks like it loses coverage of the case that `newInstance` is not overridden.

jglick replied:
I do not think we lose any coverage. In the retained original portion of the test case, the `super` call results in a nonempty `name`, so the override is a no-op. We only change the behavior in the override for the newly introduced latter half of the test case.

I mean, we lose coverage for any code calling `Util.isOverridden` on this `newInstance` overload, but I do not think there is any.

jglick commented:
> Will all of this logging of `JSONObject` result in plain logging of sensitive data, when forms contain password fields, for example?

Round-tripped password fields are encrypted. Newly entered passwords would be in plain text.

At any rate all the logging is at `FINE` or below so an administrator would need to explicitly turn it on for anything to change, and then the log files are only available to administrators, who could presumably capture passwords in other ways. If you think logging form data even at `FINE` is dangerous, I could remove the JSON from the log messages, though I did find it very helpful in analyzing how things are processed.

KostyaSha commented:
According to label this PR needs-review but was merged

jglick commented:
Yes, I tried several times without success to get review from @kohsuke, and only @daniel-beck had substantive comments. The reported bug is a regression with plenty of votes and no workaround so I did not want to leave it sitting unfixed in 1.640.

jglick commented:
(sorry, @oleg-nenashev also)

KostyaSha commented:
@jglick may it cause https://issues.jenkins-ci.org/browse/JENKINS-31954 ?

oleg-nenashev commented:
Yeah, we got into another regression. ATH should have been caught it... Or not?

amuniz commented:
This is the corresponding ATH run: https://jenkins.ci.cloudbees.com/job/core/job/acceptance-test-harness/1397

KostyaSha commented:
never saw somebody checking ATH before doing @kohsuke s releases.

jglick commented:
> ATH should have been caught it

Do you know of a particular ATH test case which would exercise the preconditions for the bug?
